### PR TITLE
updated the unit/integration templates to not have to be dependent on git

### PR DESCRIPTION
**Pull request by mbreakey3**
Now, as long as the DSCResource.Tests is already in the module (which it will be for the modules in the gallery) then there will not be any dependencies on git. Will be updating the rest of the modules as well.

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/dscresources/128)

### CODE REVIEW

**Comment by PlagueHO** [comment]
Hi there,  I haven't had a full look at this, and won't have time to comment today. You will want to make sure the version number in the template changes and the changelist.md is updated. I'm not 100% sure and this change though. @tysonjhayes, your thoughts? Ill comment later when I'm near my dev machine. :)

**Comment by mbreakey3** [comment]
Just added in the versioning

**Comment by mbreakey3** [comment]
Hold on... fixing the .md file

File: Tests.Template/changelist.md [hunk 1]
```diff
@@ -8,9 +8,13 @@ When the templates are used to create or update tests in a DSC Resource the vers
 * First release including version number.
 
 ## integration_template.ps1
+### Version 1.1.0
```
Comment (KarolKaczmarek) [inline Tests.Template/changelist.md]: If go to the "View" page you can see it doesn't get formatted properly.
Probably empty lines in between are required

**Comment by KarolKaczmarek** [comment]
@PlagueHO We want to get rid of the git dependency if DscResource.Tests is present. This is related to https://github.com/PowerShell/DscResources/issues/107 
We will also make sure new packages published to PSGallery contain DscResource.Tests.

**Comment by mbreakey3** [comment]
Okay, everything should be updated now.

File: Tests.Template/changelist.md (hunk 1 above)
Comment (mbreakey3) [inline Tests.Template/changelist.md]: Yeah, it just needed an line in between with no bullet point.

**Comment by PlagueHO** [comment]
@KarolKaczmarek, @mbreakey3 - Oh right, I complete understand and agree.

I'm mainly thinking about when I'm working on a local fork and changes have been made to DSCResource.tests since I last force updated DSCResource.tests. When I run my local tests everything will be fine, but may fail the updated tests in DSCResource.tests when pushed.

I was more thinking that changing the code to:
```powershell
if ( (-not (Test-Path -Path (Join-Path -Path $moduleRoot -ChildPath 'DSCResource.Tests'))) -or `
     (-not (Test-Path -Path (Join-Path -Path $moduleRoot -ChildPath 'DSCResource.Tests\TestHelper.psm1'))) )
{
    & git @('clone','https://github.com/PowerShell/DscResource.Tests.git',(Join-Path -Path $moduleRoot -ChildPath '\DSCResource.Tests\'))
}
elseif (-not (Get-Command 'git.exe' -ErrorAction SilentlyContinue))
{
    & git @('-C',(Join-Path -Path $moduleRoot -ChildPath '\DSCResource.Tests\'),'pull')
}
```
Would allow it to function without Git installed and still ensure DSCResource.test is the latest version when tests are run.

However, this really is such a minor thing and not exactly the end of the world should it happen - although this could be a head scratcher. :smile: I hope I'm not being mr. difficult!

**Comment by KarolKaczmarek** [comment]
I'd like to avoid pulling anything else other than what's in the official package on the production machine. 
There are other ways to detect whether module was installed from github or via psgallery, so it's still totally doable, but DscResource.tests change very rarely so I'm not sure it's worth adding it, at least for now. If we hear from many people it's an issue, I'm open to implementing this change though.

**Comment by vors** [comment]
I'm a little bit confused. Why do we want to include test code into every package in PSGallery?

**Comment by mbreakey3** [comment]
So before, when a user wanted to run a test, the DSCResource.Tests folder would be pulled from gitHub, but we don't want the users to need to have gitHub on their machine to run tests from modules that they've downloaded from the gallery, so now we're going to add in the DSCResource.Tests folder to the module before publishing to the gallery (but leave it out of the gitHub repo)

**Comment by KarolKaczmarek** [comment]
@vors Take a look at https://github.com/PowerShell/DscResources/issues/107

**Comment by vors** [comment]
Ok, I see.

Let's take a step back. I feel that our incremental steps may lead us in a wrong direction.
I'm not saying they are, but I think it's worth time to talk about it.

I want to provide the overview of `DSCResource.Tests`

### Original thinking
There was one job they accomplished: make sure that people don't send git changes that re-writes the whole file. There are two main sources of such problems:
- Encodings (so we unified on UTF-8 without BOM)
- Tab/spaces (we choose 4 spaces for tabs)

This goal was completely git-specific, so my approach was: drop these tests on CI (in `appveyor.yml`) somewhere and make `Invoke-Pester` execute them.
There was **no goal of letting users run tests locally**.

### Evolution

It turns out to be pretty annoying: contributor got all tests passed on her machine, then she send a PR just to find out that some mystical tests failing. So we put more instructions and docs about this tests. And automation.

Then we found that it would be convenient to run more common tests (i.e. DscResourceDesigner to verify `mof` schema) and so on. And common test suite grew.
Now, it's clearly desirable to run them locally, because they are non trivial (even if they been at the beginning).

### What I don't like about current state

- Running `git clone` inside `Invoke-Pester` is truly bad. Sure, we need to get rid of it.
- Shipping common test suite (that is sort of github specific) in every-module feels wrong. Just think about the amount of duplicated unused code on all production machines. Plus it doesn't solve module-developer scenario: they still need to get this tests to the machine somehow.

### Alternative solution
Maybe we should pack `DSCResource.Tests` as a module.
Then we can check from tests, that it's available and print out warning, if it's not.
In case of CI (check for specific `$env` variable), we would turn this warning into an error.
This way, we achieve
- easy local-run experience for tests. Warning message will describe what's going on and so on.
- running tests in CI
- avoid duplication of code in the gallery.

**Comment by vors** [comment]
@KarolKaczmarek @PlagueHO sorry, I missed the original conversation in #107.

**Comment by PlagueHO** [comment]
@mbreakey3, @KarolKaczmarek - sure thing, I completely agree - getting the module from PSGallery should not require Git to run the full suite of tests. I just wanted to see if we could meet both that requirement and my "I often forget to update my local DSCResource.Tests on each resource repo I'm working on" requirement :smile:

One other benefit of your approach that hasn't been mentioned here is that the DSCResource.tests included with the module in the PSGallery will always be the ones that were actually run against this module when it was released - rather than the current version which which might be newer and might actually fail when run against the currently released version of the resource.

@vors - I very much like your proposal about possibly making this a resource module. Potentially this could be a pretty good solution.

I'm only looking at things from a fairly limited point of what I need to do on a day to day basis and don't have visibility into the full release process so I'd definitely bow to @KarolKaczmarek and @mbreakey3 knowledge on this stuff.

**Comment by KarolKaczmarek** [comment]
Let's continue the discussion here: https://github.com/PowerShell/DscResources/issues/107

**Pull request merged**